protocols/v2/roles-logic-sv2/src/channel_logic/channel_factory.rs

```diff
@@ -52,12 +52,59 @@ pub enum OnNewShare {
     RelaySubmitShareUpstream,
     /// Indicate that the share meet bitcoin target, when there is an upstream the we should send
     /// the share upstream, whenever possible we should also notify the TP about it.
+    /// (share, template id, coinbase)
     ShareMeetBitcoinTarget((Share, u64, Vec<u8>)),
     /// Indicate that the share meet downstream target, in the case we could send a success
     /// response dowmstream.
     ShareMeetDownstreamTarget,
 }
 
+impl OnNewShare {
+    pub fn into_extended(&mut self, extranonce: Vec<u8>, up_id: u32) {
+        match self {
+            OnNewShare::SendErrorDownstream(_) => (),
+            OnNewShare::SendSubmitShareUpstream(share) => match share {
+                Share::Extended(_) => (),
+                Share::Standard((share, _)) => {
+                    let share = SubmitSharesExtended {
+                        channel_id: up_id,
+                        sequence_number: share.sequence_number,
+                        job_id: share.job_id,
+                        nonce: share.nonce,
+                        ntime: share.ntime,
+                        version: share.version,
+                        extranonce: extranonce.try_into().unwrap(),
+                    };
+                    let mut self_ = Self::SendSubmitShareUpstream(Share::Extended(share));
```

Comment:
`*self = Self::Send...` might be easier to read, unless I missed a reason to do it this way (same on lines 95 - 100).